### PR TITLE
Adding sc-dashboard

### PR DESCRIPTION
**Pull request by GabrielAguirre1**
Adding Home.py and test_reports.py

### CODE REVIEW

File: siliconcompiler/apps/sc_dashboard.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2023 Silicon Compiler Authors. All Rights Reserved.
+import sys
+import siliconcompiler
+
+
+def main():
+    progname = "sc-dashboard"
+    description = """
+    --------------------------------------------------------------
+    Restricted SC app that opens a dashboard...
```
Comment (gadfort) [inline siliconcompiler/apps/sc_dashboard.py:10]: Need to fix this:
Should probably be expanded to say "... a dashboard with the provided manifest" along with Examples for how to launch.
```
>>> sc-dashboard -cfg <build>
```

File: siliconcompiler/report/Home.py [hunk 2]
```diff
@@ -0,0 +1,601 @@
+import streamlit
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:1]: rename to `streamlit_viewer.py` or something like it.

File: siliconcompiler/report/Home.py [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+
+success_color = '#8EA604'  # green
+pending_color = '#F5BB00'  # yellow, could use: #EC9F05
+failure_color = '#FF4E00'  # red
+
+inactive_toggle_color = "#D3D3D3"
+active_toggle_color = "#11567f"
+track_toggle_color = "#29B5E8"
+
+# TODO: change to accomodate different operating systems
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:23]: remove comment. this will work on all.
if you wanted to you could change this to `os.path.join(os.path.dirname(__file__), '..', 'data', 'logo.png')`

File: siliconcompiler/report/Home.py [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+    "https://github.com/siliconcompiler/siliconcompiler/"
+]
+
+streamlit.set_page_config(page_title="SiliconCompiler",
+                          page_icon=Image.open(sc_logo_path),
+                          layout="wide",
+                          menu_items={
+                              "Get help": "https://docs.siliconcompiler.com/",
+                              "Report a Bug": '''https://github.com/siliconcomp
+                              iler/siliconcompiler/issues''',
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:40]: this should probably be reformatted to keep the URL together. You could do a newline at `page_title` to avoid as much white space.

File: siliconcompiler/report/Home.py [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+
+sc_about = [
+    f"SiliconCompiler {sc_version}",
+    '''A compiler framework that automates translation from source code to
+     silicon.''',
+    "https://www.siliconcompiler.com/",
+    "https://github.com/siliconcompiler/siliconcompiler/"
+]
+
+streamlit.set_page_config(page_title="SiliconCompiler",
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:34]: would be nice if the title was `{chip.design} dashboard`

File: siliconcompiler/report/Home.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+                              "Get help": "https://docs.siliconcompiler.com/",
+                              "Report a Bug": '''https://github.com/siliconcomp
+                              iler/siliconcompiler/issues''',
+                              "About": "\n\n".join(sc_about)
+                          })
+
+
+def modify_logs_and_reports_to_streamlit(logs_and_reports):
+    """
+    Converts the logs_and_reports found to the strucutre
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:47]: strucutre -> structure

File: siliconcompiler/report/Home.py [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+                          layout="wide",
+                          menu_items={
+                              "Get help": "https://docs.siliconcompiler.com/",
+                              "Report a Bug": '''https://github.com/siliconcomp
+                              iler/siliconcompiler/issues''',
+                              "About": "\n\n".join(sc_about)
+                          })
+
+
+def modify_logs_and_reports_to_streamlit(logs_and_reports):
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:45]: maybe this should be called `_convert_filepaths()`. The to_streamlit part of the name is redundent

File: siliconcompiler/report/Home.py [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+    required by streamlit_tree_select. Success is predicated on the order of
+    logs_and_reports outlined in report.get_logs_and_reports
+
+    Args:
+        logs_and_reports (list) : a list of 3-tuples with order of a path name,
+                                  folder in the subdirectory, and files in the
+                                  subdirectory
+    """
+    subsect_logs_and_reports = {}
+    if logs_and_reports == []:
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:57]: `if not logs_and_reports:`

File: siliconcompiler/report/Home.py [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+    # reverse the list to start building the tree from the leaves up
+    for path_name, folders, files in reversed(logs_and_reports):
+        children = []
+        for folder in folders:
+            # assert (folder in subsect_logs_and_reports)
+            children.append(subsect_logs_and_reports[folder])
+        for file in files:
+            node = {}
+            node['label'] = file
+            # TODO: change to accomodate different operating systems
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:69]: remove this will work just fine

File: siliconcompiler/report/Home.py [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+
+    Args:
+        logs_and_reports (list) : a list of 3-tuples with order of a path name,
+                                  folder in the subdirectory, and files in the
+                                  subdirectory
+    """
+    subsect_logs_and_reports = {}
+    if logs_and_reports == []:
+        return []
+    starting_path_name = logs_and_reports[0][0]
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:59]: is this used?

File: siliconcompiler/report/Home.py [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+            node['label'] = file
+            # TODO: change to accomodate different operating systems
+            node['value'] = f'{path_name}/{file}'
+            children.append(node)
+        if starting_path_name == path_name:
+            return children
+        else:
+            node = {}
+            # TODO: change to accomodate different operating systems
+            folder = path_name[path_name.rfind('/') + 1:]
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:77]: you can use `Pathlib` to split the paths

File: siliconcompiler/report/Home.py [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+        chip (Chip) : the chip object that contains the schema read from
+        node_dependencies (dict) : dictionary where the values of the keys are
+                                   the edges
+        successful_path (set) : contains all the nodes that are part of the
+                                'winning' path
+    """
+    nodes = []
+    edges = []
+    opacity = 0.2
+    width = 1
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:98]: make these arguments with default values

File: siliconcompiler/report/Home.py [hunk 13]
```diff
@@ -0,0 +1,601 @@
...
+    for step, index in node_dependencies:
+        if (step, index) in successful_path:
+            opacity = 1
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:101]: magic number, make this an input with default

File: siliconcompiler/report/Home.py [hunk 14]
```diff
@@ -0,0 +1,601 @@
...
+            if (step, index) in chip._get_flowgraph_exit_nodes() or \
+               (step, index) in chip._get_flowgraph_entry_nodes():
+                width = 3
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:104]: magic number, make this an input with a default (what is the unit?)

File: siliconcompiler/report/Home.py [hunk 15]
```diff
@@ -0,0 +1,601 @@
...
+        if chip._is_builtin(tool, task):
+            label = step + index + "\n" + tool
+
+        nodes.append(Node(id=step + index,
+                          label=label,
+                          color=node_color,
+                          opacity=opacity,
+                          borderWidth=width,
+                          shape='oval'))
+        width = 3
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:126]: magic number

File: siliconcompiler/report/Home.py [hunk 16]
```diff
@@ -0,0 +1,601 @@
...
+        for source_step, source_index in node_dependencies[step, index]:
+            if (source_step, source_index) in successful_path and \
+               (source_step, source_index) in successful_path:
+                width = 5
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:130]: magic number

File: siliconcompiler/report/Home.py [hunk 17]
```diff
@@ -0,0 +1,601 @@
...
+            node_color = success_color
+        elif task_status == TaskStatus.ERROR:
+            node_color = failure_color
+        else:
+            node_color = pending_color
+
+        tool, task = chip._get_tool_task(step, index)
+        label = step + index + "\n" + tool + "/" + task
+        if chip._is_builtin(tool, task):
+            label = step + index + "\n" + tool
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:118]: since step+index appears in a bunch of places, you might want to do `node_name = f'{step}{index}` to simplify the code a bit

File: siliconcompiler/report/Home.py [hunk 18]
```diff
@@ -0,0 +1,601 @@
...
+    Displays the file_preview if present. If not, displays an error message.
+
+    Args:
+        display_file_content (bool) : boolean representing if we should (true)
+                                      or should not (false) display content
+    """
+    if display_file_content:
+        path = streamlit.session_state['selected'][0]
+        file_name, file_extension = os.path.splitext(path)
+        header_col, download_col = streamlit.columns([0.89, 0.11], gap='small')
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:152]: magic numbers, how did they come about and should we make those inputs?

File: siliconcompiler/report/Home.py [hunk 19]
```diff
@@ -0,0 +1,601 @@
...
+        with header_col:
+            streamlit.header('File Preview')
+        with download_col:
+            streamlit.markdown(' ')  # aligns download button with title
+            streamlit.download_button(label="Download file",
+                                      data=path,
+                                      file_name=path[path.rfind("/"):])
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:159]: use `os.path.basename()`

File: siliconcompiler/report/Home.py [hunk 20]
```diff
@@ -0,0 +1,601 @@
...
+
+        if file_extension == ".png":
+            streamlit.image(path)
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:162]: maybe make this `if file_extension.lower() in ('.png', '.jpg', etc...):`

File: siliconcompiler/report/Home.py [hunk 21]
```diff
@@ -0,0 +1,601 @@
...
+        else:
+            with open(path, 'r') as file:
+                content = file.read()
+            if file_extension == ".json":
+                streamlit.json(content)
+            else:
+                streamlit.code(content, language='markdwon', line_numbers=True)
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:169]: markdwon? what are the options? should we expand this fully? by extensions?

File: siliconcompiler/report/Home.py [hunk 22]
```diff
@@ -0,0 +1,601 @@
...
+            streamlit.download_button(label="Download file",
+                                      data=path,
+                                      file_name=path[path.rfind("/"):])
+
+        if file_extension == ".png":
+            streamlit.image(path)
+        else:
+            with open(path, 'r') as file:
+                content = file.read()
+            if file_extension == ".json":
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:166]: `file_extension.lower()`

File: siliconcompiler/report/Home.py [hunk 23]
```diff
@@ -0,0 +1,601 @@
...
+            streamlit.markdown(' ')  # aligns download button with title
+            streamlit.download_button(label="Download file",
+                                      data=path,
+                                      file_name=path[path.rfind("/"):])
+
+        if file_extension == ".png":
+            streamlit.image(path)
+        else:
+            with open(path, 'r') as file:
+                content = file.read()
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:165]: we'll need to make this support compressed files.

File: siliconcompiler/report/Home.py [hunk 24]
```diff
@@ -0,0 +1,601 @@
...
+            if file_extension == ".json":
+                streamlit.json(content)
+            else:
+                streamlit.code(content, language='markdwon', line_numbers=True)
+    else:
+        streamlit.error('Select a file in the metrics tab first!')
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:171]: could this contain a link to the tab or is that not possible?

File: siliconcompiler/report/Home.py [hunk 25]
```diff
@@ -0,0 +1,601 @@
...
+
+
+def show_logs_reports(chip, step, index):
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:174]: probably just say `show_files`

File: siliconcompiler/report/Home.py [hunk 26]
```diff
@@ -0,0 +1,601 @@
...
+        step (string) : step of node
+        index (string) : index of node
+    """
+    streamlit.caption('files')
+
+    logs_and_reports = report.get_logs_and_reports(chip, step, index)
+    logs_and_reports = modify_logs_and_reports_to_streamlit(logs_and_reports)
+
+    if logs_and_reports == []:
+        streamlit.markdown('No files to show :(')
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:189]: remove `:(`

File: siliconcompiler/report/Home.py [hunk 27]
```diff
@@ -0,0 +1,601 @@
...
+        return False
+
+    # kinda janky at the moment, does not always flip immediately
+    # to do: make so that selection changes on first click
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:193]: TODO:

File: siliconcompiler/report/Home.py [hunk 28]
```diff
@@ -0,0 +1,601 @@
...
+        streamlit.session_state['expanded'] = selected["expanded"]
+        streamlit.session_state['right after rerun'] = True
+        streamlit.experimental_rerun()
+
+    if streamlit.session_state.selected != []:
+        return True
+    return False
+
+
+def show_metrics_of_log_or_report(chip, step, index):
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:224]: show_metrics_for_file()

File: siliconcompiler/report/Home.py [hunk 29]
```diff
@@ -0,0 +1,601 @@
...
+def show_manifest(manifest):
+    """
+    Displays the manifest and a way to search through the manifest
+
+    Args:
+        manifest (dict) : represents the manifest json
+    """
+    streamlit.header('Manifest Tree')
+
+    search_cols = streamlit.columns([0.5, 0.5], gap="large")
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:254]: magic numbers

File: siliconcompiler/report/Home.py [hunk 30]
```diff
@@ -0,0 +1,601 @@
...
+        key = streamlit.text_input('Search Keys', '', placeholder="Keys")
+        if key != '':
+            manifest = report.search_manifest(manifest, key_search=key)
+
+    with value_search_col:
+        value = streamlit.text_input('Search Values', '', placeholder="Values")
+        if value != '':
+            manifest = report.search_manifest(manifest, value_search=value)
+
+    numOfKeys = report.get_total_manifest_parameter_count(manifest)
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:267]: `expand = report.get_total_manifest_parameter_count(manifest) < 20`
maybe make the 20 an input with a default

File: siliconcompiler/report/Home.py [hunk 31]
```diff
@@ -0,0 +1,601 @@
...
+                                              display_dataframe,
+                                              [])
+                options['cols'] = nodes
+
+                metrics = streamlit.multiselect('Which metrics to include?',
+                                                displayOptions,
+                                                [])
+                options['rows'] = []
+                for metric in metrics:
+                    options['rows'].append(displayToData[metric])
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:357]: seems the only difference is options['rows/cols'] so maybe this canbe simplified to:
```
if transpose:
  key0 = 'rows'
  key1 = 'cols'
else:
  key0 = 'cols'
....
```

File: siliconcompiler/report/Home.py [hunk 32]
```diff
@@ -0,0 +1,601 @@
...
+
+            streamlit.form_submit_button("Run")
+
+    if options['cols'] == [] or options['rows'] == []:
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:361]: is not options['cols'] or not options['rows']:

File: siliconcompiler/report/Home.py [hunk 33]
```diff
@@ -0,0 +1,601 @@
...
+                                                  options['cols']]),
+                            use_container_width=True)
+
+
+def show_dataframe_header():
+    """
+    Displays the header and toggle for the dataframe. If the toggle is flipped,
+    it will update the view of the dataframe accordingly.
+    """
+    headerCol, transposeCol = streamlit.columns([0.7, 0.3], gap="large")
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:383]: magic numbers, make them inputs

File: siliconcompiler/report/Home.py [hunk 34]
```diff
@@ -0,0 +1,601 @@
...
+        streamlit.session_state['transpose'] = transpose
+
+
+def show_flowgraph():
+    """
+    Displays the header and toggle for the flowgraph, and the flowgraph itself.
+    This function shows the flowgraph. If the toggle is flipped, the flowgraph
+    will disappear.
+    """
+    home_tab_cols = streamlit.columns([0.4, 0.6], gap="large")
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:416]: magic numbers

File: siliconcompiler/report/Home.py [hunk 35]
```diff
@@ -0,0 +1,601 @@
...
+    flowgraph_col, metrics_and_nodes_info_col = home_tab_cols
+
+    with flowgraph_col:
+        headerCol, toggleCol = streamlit.columns([0.7, 0.3], gap="large")
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:420]: magic numbers

File: siliconcompiler/report/Home.py [hunk 36]
```diff
@@ -0,0 +1,601 @@
...
+    return node_from_flowgraph, metrics_and_nodes_info_col
+
+
+def dont_show_flowgraph():
+    """
+    Displays the header and toggle for the flowgraph, and the flowgraph itself.
+    This function doesn't show the flowgraph. If the toggle is flipped, the
+    flowgraph will re-appear.
+    """
+    home_tab_cols = streamlit.columns([0.1, 0.9], gap="large")
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:463]: magic numbers

File: siliconcompiler/report/Home.py [hunk 37]
```diff
@@ -0,0 +1,601 @@
...
+    flowgraph_col, metrics_and_nodes_info_col = home_tab_cols
+
+    with flowgraph_col:
+        streamlit.markdown("\n")
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:467]: why?

File: siliconcompiler/report/Home.py [hunk 38]
```diff
@@ -0,0 +1,601 @@
...
+
+
+parser = argparse.ArgumentParser('dashboard')
+parser.add_argument('cfg', nargs='?')
+args = parser.parse_args()
+
+if not args.cfg:
+    raise ValueError('configuration not provided')
+
+title_col, job_select_col = streamlit.columns([0.7, 0.3], gap="large")
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:492]: magic numbers

File: siliconcompiler/report/Home.py [hunk 39]
```diff
@@ -0,0 +1,601 @@
...
+    new_chip = Chip(design='')
+    job = streamlit.session_state['job']
+    if job == 'default':
+        new_chip = chip
+    else:
+        new_chip.schema = chip.schema.history(job)
+        new_chip.set('design', chip.design)
+
+with title_col:
+    streamlit.title(f'{new_chip.design} dashboard')
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:514]: could we add the SC logo here? lets make that another PR.

File: siliconcompiler/report/Home.py [hunk 40]
```diff
@@ -0,0 +1,601 @@
...
+    tabs = streamlit.tabs(["Metrics",
+                           "Manifest",
+                           "File Preview",
+                           "Design Preview"])
+    metrics_tab, manifest_tab, file_preview_tab, design_preview_tab = tabs
+
+    with design_preview_tab:
+        streamlit.header('Design Preview')
+
+        streamlit.image(f'{new_chip._getworkdir()}/{new_chip.design}.png')
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:555]: what happens if this is missing? maybe check if this file is there?

File: siliconcompiler/report/Home.py [hunk 41]
```diff
@@ -0,0 +1,601 @@
...
+
+    with datafram_and_node_info_col:
+        show_dataframe_header()
+
+        show_dataframe_and_parameter_selection(metric_dataframe)
+
+        streamlit.header('Node Information')
+
+        node_info_cols = streamlit.columns(3, gap='small')
+        metrics_col, records_col, logs_and_reports_col = node_info_cols
```
Comment (gadfort) [inline siliconcompiler/report/Home.py:574]:
```
metrics_col, records_col, logs_and_reports_col = \
  streamlit.columns(3, gap='small')
```

File: siliconcompiler/report/report.py [hunk 42]
```diff
@@ -0,0 +1,402 @@
...
+                           metric, step=step, index=index)
+        for source in sources:
+            if source in file_to_metric:
+                file_to_metric[source].append(metric)
+            else:
+                file_to_metric[source] = [metric]
+    return file_to_metric
+
+
+def get_logs_and_reports(chip, step, index):
```
Comment (gadfort) [inline siliconcompiler/report/report.py:356]: `get_files()`

File: siliconcompiler/report/report.py [hunk 43]
```diff
@@ -0,0 +1,402 @@
...
+        step (string) : step of node
+        index (string) : index of node
+    """
+    # could combine filters, but slighlty more efficient to seperate them
+    filters = {"inputs",
+               "reports",
+               "outputs",
+               f'{step}.log',
+               f'{step}.errors',
+               f'{step}.warnings'}
```
Comment (gadfort) [inline siliconcompiler/report/report.py:373]: if the filter is this broad, maybe we should just skip the filter and display all files?

File: siliconcompiler/report/streamlit_report.py [hunk 44]
```diff
@@ -0,0 +1,113 @@
...
+import multiprocessing
+import subprocess
+import atexit
+import shutil
+
+
+class Dashboard():
+    def __init__(self, chip, port=8501):
+        if not _has_streamlit:
+            raise RuntimeError("Streamlit is not permitted on Python 3.6")
```
Comment (gadfort) [inline siliconcompiler/report/streamlit_report.py:22]: I think this is wrong, I think it's that they moved some things around and we can fix in this PR for 3.6. I'll handle that.

File: tests/core/test_reports.py [hunk 45]
```diff
@@ -0,0 +1,317 @@
+from siliconcompiler.report import report
+from siliconcompiler import core
```
Comment (gadfort) [inline tests/core/test_reports.py:2]: import Chip

**Review by gadfort — CHANGES_REQUESTED** [review]
Looks good, but needs some cleanup in Home.py.
please aim to avoid magic numbers and give them names that relate to what they are.

File: siliconcompiler/report/streamlit_report.py [hunk 46]
```diff
@@ -0,0 +1,113 @@
...
+import time
+import tempfile
+import json
+
+try:
+    from streamlit.web import bootstrap
+    from streamlit import config as _config
+    _has_streamlit = True
+except ModuleNotFoundError:
+    _has_streamlit = False
```
Comment (troibe) [inline siliconcompiler/report/streamlit_report.py:11]: This seems counterproductive to what you are trying to achieve.
If I try to attach `chip._dashboard()` to the `heartbeat.py` example this leads to no `ModuleNotFoundError` being thrown.
Instead I'd suggest to just rely on Python's built in way of handling modules.
```suggestion
from streamlit.web import bootstrap
from streamlit import config as _config
```
That way I actually got this error as you originally intended I assume:
`ModuleNotFoundError: No module named 'streamlit'`

**Comment by troibe** [comment]
Having the units in the second column seems a bit odd but I'm not sure if streamlit's tables offer any other options...

File: siliconcompiler/report/Home.py (hunk 10 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:59]: Yes, it is used in line 72

File: siliconcompiler/report/Home.py (hunk 18 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:152]: They probably should not be inputs. All widgets are left-aligned and I found that when I made download_col the right 11% of the screen, the download button was the farthest right it could go

File: siliconcompiler/report/Home.py (hunk 21 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:169]: Yes, we definitely could expand this to all other options.

File: siliconcompiler/report/Home.py (hunk 24 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:171]: That's not possible from what I've seen, unfortunately. There is a streamlit-extras widget that lets you programmatically change between tabs, but there are some drawbacks to that as well.

File: siliconcompiler/report/Home.py (hunk 18 above)
Comment (gadfort) [reply siliconcompiler/report/Home.py:152]: how does this scale with screensize then?

File: siliconcompiler/report/Home.py (hunk 37 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:467]: This is akin to the magic numbers. I thought that it looked better this way. I will add a comment to make that clearer.

File: siliconcompiler/report/Home.py (hunk 40 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:555]: I do that a few lines earlier.

File: siliconcompiler/report/report.py (hunk 43 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/report.py:373]: Sure

**Comment by GabrielAguirre1** [comment]
> Is there some way to automatically switch to the file preview when you click on a file in the `files` section? Do the checkboxes mean it would be possible to open more than one file in the previewer?

There is no radio format for this widget, so the checkboxes are something we have to deal with. Currently, you can only select one file, but I could certainly change it such that users could select many.

**Comment by GabrielAguirre1** [comment]
> Having the units in the second column seems a bit odd but I'm not sure if streamlit's tables offer any other options...

I'm not really sure where else it would go. Do you have any suggestions?

File: siliconcompiler/report/Home.py (hunk 14 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:104]: I did the first part. The documentation really isn't clear on what the unit is.

File: siliconcompiler/report/Home.py (hunk 16 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:130]: I've oscillated between making these inputs with defaults versus constants. I'm leaning towards making them inputs with defaults since these constants are only used in this function.

**Comment by troibe** [comment]
> > Having the units in the second column seems a bit odd but I'm not sure if streamlit's tables offer any other options...
> 
> I'm not really sure where else it would go. Do you have any suggestions?

Seems like the most common thing is to either put it in parantheses in the row header or inside the cells.
https://ux.stackexchange.com/questions/78835/how-best-to-label-a-column-that-contains-values-of-different-units
But it's more of a design decision so decide for yourself if you want to change it or leave it.

File: siliconcompiler/report/Home.py (hunk 41 above)
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py:574]: I did this, but this is essentially the same as the magic numbers above. Should I treat this differently?

**Comment by GabrielAguirre1** [comment]
> > > Having the units in the second column seems a bit odd but I'm not sure if streamlit's tables offer any other options...
> > 
> > 
> > I'm not really sure where else it would go. Do you have any suggestions?
> 
> Seems like the most common thing is to either put it in parantheses in the row header or inside the cells. https://ux.stackexchange.com/questions/78835/how-best-to-label-a-column-that-contains-values-of-different-units But it's more of a design decision so decide for yourself if you want to change it or leave it.

This will work very well when the dataframe is transposed, so I think I will change it. Thanks for the suggestion.

**Review by nmoroze — COMMENTED** [review]
Nice work, this is awesome! I tried cloning this branch and experimenting with the dashboard on a run myself and it looked great. 

I did notice a bunch of dictionaries being printed out as I was interacting with the dashboard, so there might be a debug print that snuck in somewhere... it would be good to take care of that.

Most of my inline feedback is code style nitpick type things, shouldn't be anything too major!

File: siliconcompiler/report/Home.py [hunk 47]
```diff
@@ -0,0 +1,601 @@
...
+    """
+    subsect_logs_and_reports = {}
+    if logs_and_reports == []:
+        return []
+    starting_path_name = logs_and_reports[0][0]
+    # reverse the list to start building the tree from the leaves up
+    for path_name, folders, files in reversed(logs_and_reports):
+        children = []
+        for folder in folders:
+            # assert (folder in subsect_logs_and_reports)
```
[the remark this answers is not in the document]
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py]: No, I wasn't planning on it. I will delete it.

File: siliconcompiler/report/Home.py [hunk 48]
```diff
@@ -0,0 +1,601 @@
...
+                           expanded=streamlit.session_state['expanded'],
+                           only_leaf_checkboxes=True)
+
+    if len(selected['checked']) == 0:
+        streamlit.session_state['selected'] = []
+    if len(selected["checked"]) == 1:
+        streamlit.session_state['selected'] = selected["checked"]
+    if len(selected["checked"]) > 1:
+        for x in selected["checked"]:
+            if x != streamlit.session_state['selected'][0]:
```
[the remark this answers is not in the document]
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py]: Yes, I know that one of the elements will be in `streamlit.session_state['selected']`

File: siliconcompiler/report/Home.py [hunk 49]
```diff
@@ -0,0 +1,601 @@
...
+    transpose = streamlit.session_state['transpose']
+
+    if transpose:
+        # put data back to normal if need be
+        metric_dataframe = metric_dataframe.transpose()
+
+    displayToData = {}
+    displayOptions = []
+
+    if transpose:
```
[the remark this answers is not in the document]
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py]: Sounds good.

File: siliconcompiler/report/Home.py [hunk 50]
```diff
@@ -0,0 +1,601 @@
...
+        for metric, unit in metric_dataframe.columns.tolist():
+            displayToData[metric] = (metric, unit)
+            displayOptions.append(metric)
+    else:
+        for metric, unit in metric_dataframe.index.tolist():
+            displayToData[metric] = (metric, unit)
+            displayOptions.append(metric)
+
+    options = {'cols': metric_dataframe.columns.tolist(),
+               'rows': metric_dataframe.index.tolist()}
```
[the remark this answers is not in the document]
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py]: Yes, that's right. They don't necessarily get wiped out, but if either value is `[]`, then they will. I will change it to that.

File: siliconcompiler/report/Home.py [hunk 51]
```diff
@@ -0,0 +1,601 @@
...
+                                     label_after=True,
+                                     # the colors are optional
+                                     inactive_color=inactive_toggle_color,
+                                     active_color=active_toggle_color,
+                                     track_color=track_toggle_color)
+
+        streamlit.session_state['flowgraph'] = fg_toggle
+        if streamlit.session_state['flowgraph']:
+            streamlit.session_state['right after rerun'] = True
+            streamlit.experimental_rerun()
```
[the remark this answers is not in the document]
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py]: I don't understand the reasoning behind that. Would it be cleaner to hide the use of `st_toggle_switch`?

File: siliconcompiler/report/report.py [hunk 52]
```diff
@@ -0,0 +1,402 @@
...
+    if key_search:
+        return_manifest = search_manifest_keys(return_manifest, key_search)
+    if value_search:
+        return_manifest = search_manifest_values(return_manifest, value_search)
+    return return_manifest
+
+
+def get_total_manifest_parameter_count(manifest):
+    """
+    Returns (int) the number of folders and files
```
[the remark this answers is not in the document]
Comment (GabrielAguirre1) [reply siliconcompiler/report/report.py:317]: I think this does what I intend it to do, but my description is definitely off the mark. I want it to count every key in the json manifest that I give it. I used folders/files as a proxy for that, but it's clear that's not right.

File: siliconcompiler/report/report.py [hunk 53]
```diff
@@ -0,0 +1,402 @@
...
+                    metric_unit = None
+                    value = units.format_time(value)
+                else:
+                    value = units.format_si(value, metric_unit)
+
+                metrics[step, index][metric] = str(value)
+
+        if show_metric:
+            metrics_to_show.append(metric)
+            metrics_unit[metric] = metric_unit if metric_unit else ''
```
[the remark this answers is not in the document]
Comment (gadfort) [reply siliconcompiler/report/report.py:87]: We're going to handle that in a separate PR

File: siliconcompiler/report/Home.py (hunk 51 above)
[the remark this answers is not in the document]
Comment (GabrielAguirre1) [reply siliconcompiler/report/Home.py]: All set

**Review by gadfort — APPROVED** [review]
Looks good to me
